File: web/src/dashboard/relativeTime.ts
/** "just now" / "Nm ago" / "Nh ago" / "Nd ago" from a Firestore Timestamp or epoch ms. */
export function relativeTime(createdAt: unknown): string {
  const ms =
    createdAt && typeof (createdAt as { toMillis?: () => number }).toMillis === "function"
      ? (createdAt as { toMillis: () => number }).toMillis()
      : typeof createdAt === "number"
        ? createdAt
        : null;
  if (ms === null) return "";
  const diff = Date.now() - ms;
  const min = Math.round(diff / 60000);
  if (min < 1) return "just now";
  if (min < 60) return `${min}m ago`;
  const hr = Math.round(min / 60);
  if (hr < 24) return `${hr}h ago`;
  return `${Math.round(hr / 24)}d ago`;
}

File: ios/Autoloop/Domain/RelativeTime.swift
import Foundation

/// "just now" / "Nm ago" / "Nh ago" / "Nd ago" from a Date (mirrors web relativeTime.ts).
/// Returns "" for a nil date. Rounds to the nearest unit, matching the web's Math.round.
func relativeTime(_ date: Date?, now: Date = Date()) -> String {
    guard let date else { return "" }
    let diff = now.timeIntervalSince(date)
    let min = Int((diff / 60).rounded())
    if min < 1 { return "just now" }
    if min < 60 { return "\(min)m ago" }
    let hr = Int((Double(min) / 60).rounded())
    if hr < 24 { return "\(hr)h ago" }
    return "\(Int((Double(hr) / 24).rounded()))d ago"
}

/// Clock time for a session/loop row, e.g. "2:45 PM" (locale short time).
func shortTime(_ date: Date?) -> String {
    guard let date else { return "" }
    let fmt = DateFormatter()
    fmt.timeStyle = .short
    fmt.dateStyle = .none
    return fmt.string(from: date)
}
